src/charm.py

```diff
@@ -121,92 +123,107 @@ def _get_database_relation_info(self) -> dict:
+        """Runs a command to create SQL schemas and apply migration plans."""
+        process = self._container.exec(
+            ["hydra", "migrate", "sql", "-e", "--config", self._hydra_config_path, "--yes"],
+            timeout=20.0,
+        )
         try:
-            current_config = self._container.pull(self._hydra_config_path).read()
-        except (ProtocolError, PathError) as err:
-            logger.error(str(err))
-            self.unit.status = BlockedStatus(str(err))
-        else:
-            if current_config != self._config:
-                self._container.push(self._hydra_config_path, self._config, make_dirs=True)
-                logger.info("Updated hydra config, restarting the container")
-                self._container.restart()
-
-    def _update_container(self, event) -> None:
-        """Update configs, pebble layer and run database migration."""
-        self.unit.status = MaintenanceStatus("Configuring hydra container")
+            stdout, _ = process.wait_output()
+            logger.info(f"Executing automigration: {stdout}")
+        except ExecError as err:
+            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+            self.unit.status = BlockedStatus("Database migration job failed")
 
+    def _on_hydra_pebble_ready(self, event) -> None:
+        """Event Handler for pebble ready event."""
         if not self.model.relations["pg-database"]:
             # TODO: Observe relation_joined event instead of event deferring
             event.defer()
             logger.error("Missing required relation with postgresql")
             self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
             return
 
-        if not self.database.is_database_created():
-            event.defer()
-            logger.info("Missing database details. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting for database creation")
-            return
-
         if not self._container.can_connect():
             event.defer()
             logger.info("Cannot connect to Hydra container. Deferring the event.")
             self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
-        self._update_layer()
+        self.unit.status = MaintenanceStatus("Configuring resources")
+
+        self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
+        logger.info("Pebble plan updated with new configuration, replanning")
+        try:
+            self._container.replan()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+        if self.database.is_database_created():
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            self._container.start(self._container_name)
+            self.unit.status = ActiveStatus()
```

Comment:
We are also doing this in `_on_database_created`, do we need to do it again here?

Reply:
But do you know if the config file has changed or not? Maybe we can push the config file every time.

Reply:
We don't need to check if the config was changed. If it does, it will be pushed `_on_database_changed` (for now the only thing that could change in the config is the endpoint).
So there's no need to push this file it every time.